main.go

```diff
@@ -533,6 +545,14 @@ func runDaemon(d *menderDaemon) error {
 				if ok {
 					ws.Wake()
 				}
+			} else if s == syscall.SIGUSR2 {
+				log.Debug("SIGUSR2 signal received.")
+				d.inventoryUpdate <- true
+				// If the state machine is in a wait state - force a wake-up.
+				ws, ok := d.mender.GetCurrentState().(WaitState)
+				if ok {
+					ws.Wake()
+				}
```

Comment:
I have to admit I don't follow all the state transition machinery, but is there a potential race condition here? This is a separate go routine, so can it enter the wait state between `if ok` and the previous statement? If it happened it would mean that the wait state would be detected as "not waiting", but it would in fact be waiting.

Reply:
Yes, there would be. But the window would be on the order of milliseconds, so do you really think it is necessesary to handle? The worst that can happen here as far as I can see is that you have to rerun the command with a probability of 1/10e5 or less. My reasoning:
1. State-machine enters wait-state in between the check and the if statement -> no harm done. Simply rerun.
2. The state machine leaves the wait-state just in between. A wake call will be sent, but the wait-state will never be consumed by the state-machine, as it creates a new wait-state upon the next run.

Reply:
> But the window would be on the order of milliseconds, so do you really think it is necessesary to handle?

I'm afraid I do. :-) We've already had quite a few encounters with "things that should not happen", because we have used emulated devices that slow things down. On a single core embedded device which is under load, milliseconds can easily turn into seconds, and we've seen this in practice.

Luckily, I think the fix here is quite easy: Just call `Wake()` unconditionally. There is a chance that it will be stuck in the `Wake()` call for a little bit if the client is not in a wait state, but I reckon this is fine, because:

1. This is a go routine, so it doesn't hold up anything else.
2. The client is guaranteed to return to a wait state eventually.
3. Calling `Wake()` one time too many will simply spin the loop one extra time, but won't actually do anything.

Do you agree?

Reply:
> > But the window would be on the order of milliseconds, so do you really think it is necessesary to handle?
> 
> I'm afraid I do. :-) We've already had quite a few encounters with "things that should not happen", because we have used emulated devices that slow things down. On a single core embedded device which is under load, milliseconds can easily turn into seconds, and we've seen this in practice.
> 
> Luckily, I think the fix here is quite easy: Just call `Wake()` unconditionally. There is a chance that it will be stuck in the `Wake()` call for a little bit if the client is not in a wait state, but I reckon this is fine, because:
> 
> 1. This is a go routine, so it doesn't hold up anything else.
> 2. The client is guaranteed to return to a wait state eventually.
> 3. Calling `Wake()` one time too many will simply spin the loop one extra time, but won't actually do anything.
> 
> Do you agree?

I do like this approach, and as added weight to your argument, I dislike mine.

However, and please tell me I am wrong here:

Won't the problem be that not all states have a wait method, and as such:

* The user tries to send a wake-up call, but the mender-client have not yet entered a wait-state, which means that the cast will fail, and no wake call is ever sent, hence we cannot send it one times too many. Just as this is happening, the client enters a wait-state, and the window of opportunity for weirdness to happen is still here?

Please correct me if I am wrong here. I would love not to merge this.

Reply:
Darn, you're absolutely right. Well spotted. This is surprisingly complicated.

New suggestion, which is perhaps much simpler: Make a dedicated signal channel which is global, and listen to that one as well as the wakeup channel. Then post the wakeup call on this one instead. It will always exist, so there is no race condition, and the next WaitState is guaranteed to act on it. Global variables are not great, but I think it can be justified in this case since signals are also global, and it also means that you can get rid of the mutex, which would be really nice to avoid.

What do you think?